lib/prelude.dx

```diff
@@ -236,13 +236,62 @@ data Ordering =
 
 def OToW8 (x : Ordering) : Word8 = %dataConTag x
 
+'## Monoid typeclass
+
+interface Monoid a
+  mempty : a
+  mcombine : a -> a -> a  -- can't use `<>` just for parser reasons?
+
+def (<>) [Monoid a] : a -> a -> a = mcombine
+
+instance [Monoid a] Monoid (n=>a)
+  mempty = for i. mempty
+  mcombine = \x y. for i. mcombine x.i y.i
+
+named-instance AndMonoid : Monoid Bool
+  mempty = True
+  mcombine = (&&)
+
+named-instance OrMonoid : Monoid Bool
+  mempty = False
+  mcombine = (||)
+
+def AddMonoid (a:Type) -> (_:Add a) ?=> : Monoid a =
+  A = a -- XXX: Typing `Monoid a` below would quantify it over a, which we don't want
+  named-instance result : Monoid A
+    mempty = zero
+    mcombine = add
+  result
+
+def MulMonoid (a:Type) -> (_:Mul a) ?=> : Monoid a =
+  A = a -- XXX: Typing `Monoid a` below would quantify it over a, which we don't want
+  named-instance result : Monoid A
+    mempty = one
+    mcombine = mul
+  result
+
```

Comment:
I love how instances work/are written now so much.
I think we are not quite at their final form, but wow this is so much nicer than before.